File: tokenmaxxing-macos/Sources/Tokenmaxxing/Views/Charts.swift
import SwiftUI

struct BarRow: Identifiable {
    let id = UUID()
    var label: String
    var value: Double
    var caption: String
    var color: Color
}

struct DonutSlice {
    var value: Double
    var color: Color
}

struct ResetTick: Identifiable {
    let id = UUID()
    var label: String
    var seconds: Int
    var trusted: Bool
    var color: Color
}

struct LegendItem: Identifiable {
    let id = UUID()
    var label: String
    var color: Color
    var value: String
}

/// A filled area chart of one series (daily cost or tokens), with a bright top
/// line and a highlighted last point.
struct AreaChart: View {
    var series: [Double]
    var accent: Color

    var body: some View {
        Canvas { ctx, size in
            let padY = size.height * 0.10
            let w = size.width
            let h = size.height
            let plotH = h - padY * 2

            var baseline = Path()
            baseline.move(to: CGPoint(x: 0, y: h - padY))
            baseline.addLine(to: CGPoint(x: w, y: h - padY))
            ctx.stroke(baseline, with: .color(Palette.track.opacity(0.8)), lineWidth: 1)

            guard series.count >= 2 else { return }
            let maxV = Swift.max(series.max() ?? 1, 0.000001)
            func point(_ i: Int, _ v: Double) -> CGPoint {
                CGPoint(
                    x: CGFloat(Double(i) / Double(series.count - 1)) * w,
                    y: (h - padY) - CGFloat(v / maxV) * plotH
                )
            }

            var fill = Path()
            fill.move(to: CGPoint(x: 0, y: h - padY))
            for (i, v) in series.enumerated() { fill.addLine(to: point(i, v)) }
            fill.addLine(to: CGPoint(x: w, y: h - padY))
            fill.closeSubpath()
            ctx.fill(fill, with: .linearGradient(
                Gradient(colors: [accent.opacity(0.34), accent.opacity(0.015)]),
                startPoint: CGPoint(x: 0, y: padY),
                endPoint: CGPoint(x: 0, y: h - padY)
            ))

            var line = Path()
            for (i, v) in series.enumerated() {
                let p = point(i, v)
                if i == 0 { line.move(to: p) } else { line.addLine(to: p) }
            }
            ctx.stroke(line, with: .color(accent), style: StrokeStyle(lineWidth: Swift.max(1.6, h * 0.02), lineCap: .round, lineJoin: .round))

            if let last = series.last {
                let p = point(series.count - 1, last)
                ctx.fill(Path(ellipseIn: CGRect(x: p.x - h * 0.05, y: p.y - h * 0.05, width: h * 0.1, height: h * 0.1)), with: .color(accent.opacity(0.28)))
                ctx.fill(Path(ellipseIn: CGRect(x: p.x - h * 0.03, y: p.y - h * 0.03, width: h * 0.06, height: h * 0.06)), with: .color(accent))
            }
        }
    }
}

/// A donut of proportional slices with two lines of centered text.
struct DonutChart: View {
    var slices: [DonutSlice]
    var centerTop: String
    var centerBottom: String

    var body: some View {
        Canvas { ctx, size in
            let cx = size.width / 2
            let cy = size.height / 2
            let radius = Swift.min(size.width, size.height) / 2 - size.height * 0.06
            let thickness = radius * 0.44
            let total = slices.reduce(0) { $0 + $1.value }
            let ringR = radius - thickness / 2

            var track = Path()
            track.addArc(center: CGPoint(x: cx, y: cy), radius: ringR, startAngle: .degrees(0), endAngle: .degrees(360), clockwise: false)
            ctx.stroke(track, with: .color(Palette.track), style: StrokeStyle(lineWidth: thickness))

            if total > 0 {
                var start = -90.0
                let gap = 2.5
                for s in slices where s.value > 0 {
                    let sweep = s.value / total * 360
                    var arc = Path()
                    arc.addArc(
                        center: CGPoint(x: cx, y: cy), radius: ringR,
                        startAngle: .degrees(start + gap),
                        endAngle: .degrees(start + sweep - Swift.min(gap, sweep / 2)),
                        clockwise: false
                    )
                    ctx.stroke(arc, with: .color(s.color), style: StrokeStyle(lineWidth: thickness, lineCap: .butt))
                    start += sweep
                }
            }

            ctx.draw(
                Text(centerTop).font(.system(size: radius * 0.4, weight: .bold, design: .rounded)).foregroundStyle(Palette.text),
                at: CGPoint(x: cx, y: cy - radius * 0.04), anchor: .center
            )
            ctx.draw(
                Text(centerBottom).font(.system(size: radius * 0.2)).foregroundStyle(Palette.muted),
                at: CGPoint(x: cx, y: cy + radius * 0.28), anchor: .center
            )
        }
    }
}

/// A 7×24 activity punch card: rows are weekdays (Mon top), columns are hours,
/// cell brightness scales with activity (square-root so light days stay visible).
struct HeatmapChart: View {
    var counts: [[Int]]
    var maxV: Int
    var accent: Color
    private let days = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]

    var body: some View {
        Canvas { ctx, size in
            let gutterL = size.width * 0.075
            let gutterB = size.height * 0.16
            let cellW = (size.width - gutterL) / 24
            let cellH = (size.height - gutterB) / 7
            let inset = Swift.min(Swift.min(cellW, cellH) * 0.12, 2)
            let denom = Double(Swift.max(maxV, 1))

            for row in 0..<7 {
                let cy = CGFloat(row) * cellH
                ctx.draw(
                    Text(days[row]).font(.system(size: Swift.min(Swift.max(cellH * 0.44, 8), 13))).foregroundStyle(Palette.muted),
                    at: CGPoint(x: gutterL - inset * 2, y: cy + cellH * 0.5), anchor: .trailing
                )
                for hour in 0..<24 {
                    let x = gutterL + CGFloat(hour) * cellW
                    let rect = CGRect(x: x + inset, y: cy + inset, width: cellW - inset * 2, height: cellH - inset * 2)
                    let path = Path(roundedRect: rect, cornerRadius: Swift.max(inset, 1.5))
                    let count = row < counts.count && hour < counts[row].count ? counts[row][hour] : 0
                    if count == 0 {
                        ctx.fill(path, with: .color(Palette.track.opacity(0.5)))
                    } else {
                        let intensity = Swift.min(Swift.max((Double(count) / denom).squareRoot(), 0.12), 1.0)
                        ctx.fill(path, with: .color(accent.opacity(intensity)))
                    }
                }
            }
            for hour in [0, 6, 12, 18, 23] {
                let x = gutterL + (CGFloat(hour) + 0.5) * cellW
                ctx.draw(
                    Text("\(hour)").font(.system(size: Swift.min(Swift.max(cellH * 0.4, 8), 12))).foregroundStyle(Palette.muted),
                    at: CGPoint(x: x, y: size.height - gutterB * 0.28), anchor: .bottom
                )
            }
        }
    }
}

/// A single soonest-first timeline collapsing every reset across both providers.
struct ResetHorizonChart: View {
    var ticks: [ResetTick]

    var body: some View {
        Canvas { ctx, size in
            let axisY = size.height * 0.5
            let span = 7.0 * 86_400.0

            var axis = Path()
            axis.move(to: CGPoint(x: 0, y: axisY))
            axis.addLine(to: CGPoint(x: size.width, y: axisY))
            ctx.stroke(axis, with: .color(Palette.track), lineWidth: 2)

            for d in 0...7 {
                let x = CGFloat(Double(d) / 7) * size.width
                var g = Path()
                g.move(to: CGPoint(x: x, y: axisY - 4))
                g.addLine(to: CGPoint(x: x, y: axisY + 4))
                ctx.stroke(g, with: .color(Palette.track.opacity(0.6)), lineWidth: 1)
                ctx.draw(Text("\(d)d").font(.system(size: 9.5)).foregroundStyle(Palette.muted), at: CGPoint(x: x, y: size.height - 2), anchor: .bottom)
            }

            var lastX = -CGFloat.infinity
            var below = false
            for t in ticks.sorted(by: { $0.seconds < $1.seconds }) {
                let frac = Swift.min(Swift.max(Double(t.seconds) / span, 0), 1)
                let x = CGFloat(frac) * size.width
                let dot = Path(ellipseIn: CGRect(x: x - 5, y: axisY - 5, width: 10, height: 10))
                if t.trusted {
                    ctx.fill(dot, with: .color(t.color))
                } else {
                    ctx.stroke(dot, with: .color(t.color), lineWidth: 2)
                }
                below = abs(x - lastX) < size.width * 0.16 ? !below : false
                lastX = x
                let ly = below ? axisY + 18 : axisY - 12
                let label = Self.short(t.label) + " · " + Fmt.until(t.seconds)
                ctx.draw(
                    Text(label).font(.system(size: 10, weight: .semibold)).foregroundStyle(t.color),
                    at: CGPoint(x: Swift.min(Swift.max(x, size.width * 0.14), size.width * 0.86), y: ly), anchor: .center
                )
            }
        }
    }

    private static func short(_ s: String) -> String {
        s.split(separator: "·").last.map { $0.trimmingCharacters(in: .whitespaces) } ?? s
    }
}

/// A ranked bar breakdown: label left, caption right, proportional bar below.
struct BarsChart: View {
    var rows: [BarRow]
    private var maxV: Double { Swift.max(rows.map(\.value).max() ?? 1, 0.000001) }

    var body: some View {
        VStack(spacing: 7) {
            ForEach(rows) { row in
                VStack(spacing: 3) {
                    HStack(spacing: 6) {
                        Text(row.label)
                            .font(.system(size: 12))
                            .foregroundStyle(Palette.text.opacity(0.82))
                            .lineLimit(1)
                        Spacer()
                        Text(row.caption)
                            .font(.system(size: 11.5, weight: .semibold, design: .monospaced))
                            .foregroundStyle(Palette.text)
                    }
                    GeometryReader { geo in
                        ZStack(alignment: .leading) {
                            Capsule().fill(Palette.track).frame(height: 6)
                            Capsule().fill(row.color)
                                .frame(width: Swift.max(6, geo.size.width * CGFloat(row.value / maxV)), height: 6)
                        }
                    }
                    .frame(height: 6)
                }
            }
        }
    }
}

/// A single proportional stacked bar (token composition, free/paid split).
struct CompositionBar: View {
    var segments: [DonutSlice]
    private var total: Double { segments.reduce(0) { $0 + $1.value } }

    var body: some View {
        GeometryReader { geo in
            HStack(spacing: 0) {
                ForEach(Array(segments.enumerated()), id: \.offset) { _, seg in
                    Rectangle()
                        .fill(seg.color)
                        .frame(width: total > 0 ? geo.size.width * CGFloat(seg.value / total) : 0)
                }
            }
        }
        .frame(height: 18)
        .background(Capsule().fill(Palette.track))
        .clipShape(Capsule())
    }
}

struct LegendView: View {
    var items: [LegendItem]

    var body: some View {
        VStack(alignment: .leading, spacing: 5) {
            ForEach(items) { item in
                HStack(spacing: 8) {
                    RoundedRectangle(cornerRadius: 2).fill(item.color).frame(width: 9, height: 9)
                    Text(item.label).font(.system(size: 11.5)).foregroundStyle(Palette.text.opacity(0.82)).lineLimit(1)
                    Spacer(minLength: 6)
                    Text(item.value).font(.system(size: 11, weight: .semibold, design: .monospaced)).foregroundStyle(Palette.text)
                }
            }
            Spacer(minLength: 0)
        }
    }
}

File: tokenmaxxing-kde/src/charts.rs
//! Hand-drawn Cairo chart primitives for the dashboard. Each takes an already
//! prepared data slice and paints into a `w`×`h` box anchored at the origin, so
//! the same code serves both the live `DrawingArea`s and the share-card render.

use crate::theme::{self, Rgb};
use gtk::cairo::{Context, FontSlant, FontWeight, LineCap, LinearGradient};
use std::f64::consts::PI;

pub struct BarRow {
    pub label: String,
    pub value: f64,
    pub caption: String,
    pub color: Rgb,
}

pub struct Slice {
    pub value: f64,
    pub color: Rgb,
}

/// A filled area chart of one series (daily cost or tokens), with a bright top
/// line, a highlighted last point, and a faint peak marker.
pub fn area(cr: &Context, w: f64, h: f64, series: &[f64], accent: Rgb) {
    let pad = h * 0.14;
    let plot_w = w - pad * 2.0;
    let plot_h = h - pad * 2.0;
    baseline(cr, pad, h - pad, w - pad);
    if series.len() < 2 {
        return;
    }
    let max = series.iter().cloned().fold(0.0_f64, f64::max).max(f64::MIN_POSITIVE);
    let point = |i: usize, v: f64| {
        let x = pad + (i as f64 / (series.len() - 1) as f64) * plot_w;
        let y = (h - pad) - (v / max) * plot_h;
        (x, y)
    };

    cr.new_path();
    cr.move_to(pad, h - pad);
    for (i, &v) in series.iter().enumerate() {
        let (x, y) = point(i, v);
        cr.line_to(x, y);
    }
    cr.line_to(w - pad, h - pad);
    cr.close_path();
    let fill = LinearGradient::new(0.0, pad, 0.0, h - pad);
    fill.add_color_stop_rgba(0.0, accent.0, accent.1, accent.2, 0.34);
    fill.add_color_stop_rgba(1.0, accent.0, accent.1, accent.2, 0.015);
    let _ = cr.set_source(&fill);
    let _ = cr.fill();

    cr.set_line_cap(LineCap::Round);
    cr.set_line_join(gtk::cairo::LineJoin::Round);
    cr.set_line_width((h * 0.022).max(1.4));
    set(cr, accent);
    cr.new_path();
    for (i, &v) in series.iter().enumerate() {
        let (x, y) = point(i, v);
        if i == 0 {
            cr.move_to(x, y);
        } else {
            cr.line_to(x, y);
        }
    }
    let _ = cr.stroke();

    let (lx, ly) = point(series.len() - 1, *series.last().unwrap());
    set_alpha(cr, accent, 0.28);
    cr.arc(lx, ly, h * 0.052, 0.0, 2.0 * PI);
    let _ = cr.fill();
    set(cr, accent);
    cr.arc(lx, ly, h * 0.03, 0.0, 2.0 * PI);
    let _ = cr.fill();
}

/// A top-N horizontal bar ranking: label left, caption right, proportional bar.
pub fn bars(cr: &Context, w: f64, h: f64, rows: &[BarRow]) {
    if rows.is_empty() {
        return;
    }
    let n = rows.len();
    let row_h = h / n as f64;
    let max = rows.iter().map(|r| r.value).fold(0.0_f64, f64::max).max(f64::MIN_POSITIVE);
    let bar_h = (row_h * 0.24).clamp(4.0, 10.0);
    let text_size = (row_h * 0.34).clamp(11.0, 16.0);

    for (i, row) in rows.iter().enumerate() {
        let y = i as f64 * row_h;
        let label_baseline = y + row_h * 0.42;
        font(cr, text_size, FontWeight::Normal);
        set(cr, (0.72, 0.77, 0.85));
        text_left(cr, &ellipsize(cr, &row.label, w * 0.62), 0.0, label_baseline);
        font(cr, text_size * 0.92, FontWeight::Bold);
        set(cr, theme::TEXT);
        text_right(cr, &row.caption, w, label_baseline);

        let track_y = y + row_h * 0.62;
        rounded(cr, 0.0, track_y, w, bar_h, bar_h / 2.0);
        set(cr, theme::TRACK);
        let _ = cr.fill();
        let frac = (row.value / max).clamp(0.0, 1.0);
        if frac > 0.0 {
            rounded(cr, 0.0, track_y, (w * frac).max(bar_h), bar_h, bar_h / 2.0);
            set(cr, row.color);
            let _ = cr.fill();
        }
    }
}

/// A donut of proportional slices with two lines of centered text.
pub fn donut(cr: &Context, w: f64, h: f64, slices: &[Slice], center_top: &str, center_bottom: &str) {
    let cx = w / 2.0;
    let cy = h / 2.0;
    let radius = (w.min(h) / 2.0) - h * 0.06;
    let thickness = radius * 0.44;
    let total: f64 = slices.iter().map(|s| s.value).sum();

    cr.set_line_cap(LineCap::Butt);
    cr.set_line_width(thickness);
    set(cr, theme::TRACK);
    cr.arc(cx, cy, radius - thickness / 2.0, 0.0, 2.0 * PI);
    let _ = cr.stroke();

    if total > 0.0 {
        let gap = 0.03;
        let mut start = -PI / 2.0;
        for s in slices {
            let sweep = (s.value / total) * (2.0 * PI);
            if sweep <= 0.0 {
                continue;
            }
            set(cr, s.color);
            cr.arc(cx, cy, radius - thickness / 2.0, start + gap, start + sweep - gap.min(sweep / 2.0));
            let _ = cr.stroke();
            start += sweep;
        }
    }

    font(cr, radius * 0.44, FontWeight::Bold);
    set(cr, theme::TEXT);
    centered(cr, center_top, cx, cy - radius * 0.02);
    font(cr, radius * 0.2, FontWeight::Normal);
    set(cr, theme::MUTED);
    centered(cr, center_bottom, cx, cy + radius * 0.32);
}

/// A single proportional stacked bar (token composition, free/paid split, …).
pub fn stacked_bar(cr: &Context, x: f64, y: f64, w: f64, h: f64, segments: &[Slice]) {
    let total: f64 = segments.iter().map(|s| s.value).sum();
    rounded(cr, x, y, w, h, h / 2.0);
    set(cr, theme::TRACK);
    let _ = cr.fill();
    if total <= 0.0 {
        return;
    }
    cr.save().ok();
    rounded(cr, x, y, w, h, h / 2.0);
    cr.clip();
    let mut cx = x;
    for s in segments {
        let seg_w = (s.value / total) * w;
        if seg_w <= 0.0 {
            continue;
        }
        cr.rectangle(cx, y, seg_w, h);
        set(cr, s.color);
        let _ = cr.fill();
        cx += seg_w;
    }
    cr.restore().ok();
}

const WEEKDAYS: [&str; 7] = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"];

/// A 7×24 activity punch card: rows are weekdays (Mon top), columns are hours,
/// cell brightness scales with activity (square-root so light days stay visible).
pub fn heatmap(cr: &Context, w: f64, h: f64, counts: &[[u64; 24]; 7], max: u64, accent: Rgb) {
    let gutter_l = w * 0.075;
    let gutter_b = h * 0.16;
    let grid_w = w - gutter_l;
    let grid_h = h - gutter_b;
    let cell_w = grid_w / 24.0;
    let cell_h = grid_h / 7.0;
    let inset = (cell_w.min(cell_h) * 0.12).min(2.0);
    let max = max.max(1) as f64;

    font(cr, (cell_h * 0.44).clamp(8.0, 13.0), FontWeight::Normal);
    for (row, label) in WEEKDAYS.iter().enumerate() {
        let cy = row as f64 * cell_h;
        set(cr, theme::MUTED);
        text_right(cr, label, gutter_l - inset * 2.0, cy + cell_h * 0.66);
        for hour in 0..24 {
            let x = gutter_l + hour as f64 * cell_w;
            let count = counts[row][hour];
            rounded(cr, x + inset, cy + inset, cell_w - inset * 2.0, cell_h - inset * 2.0, inset.max(1.5));
            if count == 0 {
                set_alpha(cr, theme::TRACK, 0.5);
            } else {
                let intensity = (count as f64 / max).sqrt().clamp(0.12, 1.0);
                set_alpha(cr, accent, intensity);
            }
            let _ = cr.fill();
        }
    }

    font(cr, (cell_h * 0.4).clamp(8.0, 12.0), FontWeight::Normal);
    set(cr, theme::MUTED);
    for hour in [0, 6, 12, 18, 23] {
        let x = gutter_l + (hour as f64 + 0.5) * cell_w;
        centered(cr, &format!("{hour}"), x, h - gutter_b * 0.28);
    }
}

fn baseline(cr: &Context, x0: f64, y: f64, x1: f64) {
    set_alpha(cr, theme::TRACK, 0.8);
    cr.set_line_width(1.0);
    cr.move_to(x0, y);
    cr.line_to(x1, y);
    let _ = cr.stroke();
}

fn ellipsize(cr: &Context, text: &str, max_w: f64) -> String {
    if cr.text_extents(text).map(|e| e.width()).unwrap_or(0.0) <= max_w {
        return text.to_string();
    }
    let mut out = text.to_string();
    while out.chars().count() > 1
        && cr.text_extents(&format!("{out}…")).map(|e| e.width()).unwrap_or(0.0) > max_w
    {
        out.pop();
    }
    format!("{out}…")
}

fn font(cr: &Context, size: f64, weight: FontWeight) {
    cr.select_font_face("sans-serif", FontSlant::Normal, weight);
    cr.set_font_size(size);
}

fn text_left(cr: &Context, s: &str, x: f64, baseline: f64) {
    cr.move_to(x, baseline);
    let _ = cr.show_text(s);
}

fn text_right(cr: &Context, s: &str, right: f64, baseline: f64) {
    let width = cr.text_extents(s).map(|e| e.width()).unwrap_or(0.0);
    cr.move_to(right - width, baseline);
    let _ = cr.show_text(s);
}

fn centered(cr: &Context, s: &str, cx: f64, baseline: f64) {
    let ext = match cr.text_extents(s) {
        Ok(e) => e,
        Err(_) => return,
    };
    cr.move_to(cx - ext.width() / 2.0 - ext.x_bearing(), baseline);
    let _ = cr.show_text(s);
}

fn rounded(cr: &Context, x: f64, y: f64, w: f64, h: f64, r: f64) {
    let r = r.min(w / 2.0).min(h / 2.0).max(0.0);
    cr.new_path();
    cr.arc(x + w - r, y + r, r, -PI / 2.0, 0.0);
    cr.arc(x + w - r, y + h - r, r, 0.0, PI / 2.0);
    cr.arc(x + r, y + h - r, r, PI / 2.0, PI);
    cr.arc(x + r, y + r, r, PI, 1.5 * PI);
    cr.close_path();
}

fn set(cr: &Context, c: Rgb) {
    cr.set_source_rgb(c.0, c.1, c.2);
}

fn set_alpha(cr: &Context, c: Rgb, a: f64) {
    cr.set_source_rgba(c.0, c.1, c.2, a);
}
